### PR TITLE
Add gather collective for StringTypes

### PR DESCRIPTION
**Pull request by HendersonSC**
Adds a wrapper in parallel env which works similarly to mpi_gather for futility string types..

### CODE REVIEW

File: unit_tests/testParallelEnv/testParallelEnv.f90 [hunk 1]
```diff
@@ -350,6 +342,34 @@ SUBROUTINE testMPIEnv()
...
+      ALLOCATE(tstRecv(2))
+      CALL testMPI%gatherV(tstString,0)
+      ASSERT_EQ(CHAR(tstString(1)),"","test gather at master, 1")
+      ASSERT_EQ(CHAR(tstString(2)),"blah_not_master","test gather at master, 2")
+
+      CALL testMPI%gatherV(tstString,1)
+      ASSERT_EQ(CHAR(tstString(1)),"blah_master","test gather at master, 1")
+      ASSERT_EQ(CHAR(tstString(2)),"blah_not_master","test gather at master, 2")
+    ENDIF
+  ENDIF
```
Comment (aarograh) [inline unit_tests/testParallelEnv/testParallelEnv.f90:371]: Wouldn't be terrible to also test the `nproc == 1` case as well.

Comment (aarograh) [reply unit_tests/testParallelEnv/testParallelEnv.f90:371]: I would also like to see a test for the rank-2 array as well

File: src/ParallelEnv.f90 [hunk 2]
```diff
@@ -200,6 +201,16 @@ SUBROUTINE ParEnvType_clear_absintfc(myPE)
...
         gather_SIK1_MPI_Env_type
+
+    !> @copybrief ParallelEnv::gatherV_str1D_MPI_ENV_type
+    !> @copydetails ParallelEnv::gatherV_str1D_MPI_ENV_type
+    PROCEDURE,PASS,PRIVATE :: gatherV_str1D_MPI_ENV_type
+    !> @copybrief ParallelEnv::gatherV_str2D_MPI_ENV_type
+    !> @copydetails ParallelEnv::gatherV_str2D_MPI_ENV_type
+    PROCEDURE,PASS,PRIVATE :: gatherV_str2D_MPI_ENV_type
+    !>
+    GENERIC :: gatherV => gatherV_str1D_MPI_ENV_type,gatherV_str2D_MPI_ENV_type
```
Comment (aarograh) [inline src/ParallelEnv.f90:212]: Couldn't get this under the `gather` generic?

**Review by aarograh — CHANGES_REQUESTED** [review]
Looks good overall.  Mainly would like to see a tad more testing.

Thanks for cleaning up some of the other doxygen comments too

File: src/ParallelEnv.f90 (hunk 2 above)
Comment (HendersonSC) [reply src/ParallelEnv.f90:212]: I started with the same generic, but hit a snag (probably of my own making) with the arguments in the interface. Not sure which way around gather/gatherv makes more sense in the actual name scheme.

Comment (aarograh) [reply src/ParallelEnv.f90:212]: Hmm... ok... do you remember the error?  If not, could you try it again and let me know what it was?  I've run into errors with `GENERIC` when we used deferred shape arrays (`REAL(SRK),DIMENSION(*) :: ...`), but I feel like the current `gather` procedures should work fine.

**Review by aarograh — APPROVED** [review]

**Pull request merged**